Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp

```diff
@@ -0,0 +1,248 @@
+
+RenderSVGResourceContainer::RenderSVGResourceContainer(SVGElement& element, RenderStyle&& style)
+    : RenderSVGHiddenContainer(element, WTFMove(style))
+    , m_id(element.getIdAttribute())
+{
+}
+
+RenderSVGResourceContainer::~RenderSVGResourceContainer() = default;
+
+void RenderSVGResourceContainer::layout()
+{
+    StackStats::LayoutCheckPoint layoutCheckPoint;
+
+    // Invalidate all resources if our layout changed.
+    if (selfNeedsClientInvalidation())
+        RenderSVGRoot::addResourceForClientInvalidation(this);
+
+    RenderSVGHiddenContainer::layout();
+}
+
+void RenderSVGResourceContainer::willBeDestroyed()
+{
+    SVGResourcesCache::resourceDestroyed(*this);
+
+    if (m_registered) {
+        treeScopeForSVGReferences().removeSVGResource(m_id);
+        m_registered = false;
+    }
+
+    RenderSVGHiddenContainer::willBeDestroyed();
+}
+
+void RenderSVGResourceContainer::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle)
+{
+    RenderSVGHiddenContainer::styleDidChange(diff, oldStyle);
+
+    if (!m_registered) {
+        m_registered = true;
+        registerResource();
+    }
+}
+
+void RenderSVGResourceContainer::idChanged()
+{
+    // Invalidate all our current clients.
+    removeAllClientsFromCache();
+
+    // Remove old id, that is guaranteed to be present in cache.
+    treeScopeForSVGReferences().removeSVGResource(m_id);
+    m_id = element().getIdAttribute();
+
+    registerResource();
+}
+
+void RenderSVGResourceContainer::markAllClientsForRepaint()
+{
+    markAllClientsForInvalidation(RepaintInvalidation);
+}
+
+void RenderSVGResourceContainer::markAllClientsForInvalidation(InvalidationMode mode)
+{
+    // FIXME: Style invalidation should either be a pre-layout task or this function
+    // should never get called while in layout. See webkit.org/b/208903.
+    if ((m_clients.isEmptyIgnoringNullReferences() && m_clientLayers.isEmptyIgnoringNullReferences()) || m_isInvalidating)
+        return;
+
+    SetForScope<bool> isInvalidating(m_isInvalidating, true);
+
+    bool needsLayout = mode == LayoutAndBoundariesInvalidation;
+    bool markForInvalidation = mode != ParentOnlyInvalidation;
+    auto* root = SVGRenderSupport::findTreeRootObject(*this);
+
+    for (auto& client : m_clients) {
+        // We should not mark any client outside the current root for invalidation
+        if (root != SVGRenderSupport::findTreeRootObject(client))
+            continue;
+
+        if (is<RenderSVGResourceContainer>(client)) {
+            downcast<RenderSVGResourceContainer>(client).removeAllClientsFromCache(markForInvalidation);
+            continue;
+        }
+
+        if (markForInvalidation)
+            markClientForInvalidation(client, mode);
+
+        RenderSVGResource::markForLayoutAndParentResourceInvalidation(client, needsLayout);
+    }
+
+    markAllClientLayersForInvalidation();
+}
+
+void RenderSVGResourceContainer::markAllClientLayersForInvalidation()
```

Comment:
I am not sure we should actually add all of that invalidation handling in the beginning, but rather get e.g. simple clipping running, and then revisit the whole topic regarding invalidation, when we know for sure what clients we have to serve: RenderLayers, or RenderObjects, or both, etc.

WDYT?